Sources/MessagePack/Unpack.swift

```diff
@@ -1,12 +1,112 @@
+
 import Foundation
 
+
+/// Handle the remaining data without duplicating
+final class RemainderData {
+  let data:Data
+  private let bytes:UnsafePointer<UInt8>
+  private let rawBytes:UnsafeRawPointer
+  
+  /// Current location of where we are processing
+  private var offset:Int = 0 {
+    didSet {
+      isEmpty = offset >= data.endIndex
+    }
+  }
+  
+  /// Initialize with the original data source
+  ///
+  /// - Parameter data: The data to be unpacked
+  init(_ data:Data) {
+    self.data = data
+    rawBytes = (self.data as NSData).bytes
```

Comment:
Calling `-[NSData bytes]` causes a copy of all noncontiguous subregions. I think it would be better to create some data type containing the immutable `Data` and some kind of `offset` index (like you have) but without the `bytes` and `rawBytes` parts.